File: java/revenj-servlet/src/main/java/org/revenj/server/commands/search/CountDomainObject.java
package org.revenj.server.commands.search;

import org.revenj.patterns.*;
import org.revenj.security.PermissionManager;
import org.revenj.server.CommandResult;
import org.revenj.server.ReadOnlyServerCommand;
import org.revenj.serialization.Serialization;

import java.io.IOException;
import java.lang.reflect.Type;
import java.security.Principal;
import java.util.Optional;

public class CountDomainObject implements ReadOnlyServerCommand {

	private final DomainModel domainModel;
	private final PermissionManager permissions;

	public CountDomainObject(
			DomainModel domainModel,
			PermissionManager permissions) {
		this.domainModel = domainModel;
		this.permissions = permissions;
	}

	public static final class Argument<TFormat> {
		public String Name;
		public String SpecificationName;
		public TFormat Specification;

		public Argument(String name, String specificationName, TFormat specification) {
			this.Name = name;
			this.SpecificationName = specificationName;
			this.Specification = specification;
		}

		@SuppressWarnings("unused")
		private Argument() {
		}
	}

	@Override
	public <TInput, TOutput> CommandResult<TOutput> execute(
			ServiceLocator locator,
			Serialization<TInput> input,
			Serialization<TOutput> output,
			TInput data,
			Principal principal) {
		Argument<TInput> arg;
		try {
			arg = input.deserialize(data, Argument.class, data.getClass());
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		Optional<Class<?>> manifest = domainModel.find(arg.Name);
		if (!manifest.isPresent()) {
			return CommandResult.badRequest("Unable to find specified domain object: " + arg.Name);
		}
		if (!permissions.canAccess(manifest.get(), principal)) {
			return CommandResult.forbidden(arg.Name);
		}
		final Specification specification;
		if (arg.SpecificationName != null && arg.SpecificationName.length() > 0) {
			Optional<Class<?>> specType = domainModel.find(arg.Name + '+' + arg.SpecificationName);
			if (!specType.isPresent()) {
				specType = domainModel.find(arg.SpecificationName);
			}
			if (!specType.isPresent()) {
				return CommandResult.badRequest("Couldn't find specification: " + arg.SpecificationName);
			}
			try {
				specification = (Specification) input.deserialize((Type) specType.get(), arg.Specification);
			} catch (IOException e) {
				return CommandResult.badRequest("Error deserializing specification: " + arg.SpecificationName);
			}
		} else if (arg.Specification instanceof Specification) {
			specification = (Specification) arg.Specification;
		} else {
			specification = null;
		}
		SearchableRepository repository;
		try {
			repository = locator.resolve(SearchableRepository.class, manifest.get());
		} catch (ReflectiveOperationException e) {
			return CommandResult.badRequest("Error resolving repository for: " + arg.Name + ". Reason: " + e.getMessage());
		}
		long found = repository.count(specification);
		try {
			return CommandResult.success(Long.toString(found), output.serialize(found));
		} catch (IOException e) {
			return new CommandResult<>(null, "Error serializing result.", 500);
		}
	}
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/commands/search/CountDomainObject.scala
package net.revenj.server.commands.search

import net.revenj.patterns._
import net.revenj.serialization.Serialization
import net.revenj.server.commands.search.CountDomainObject.Argument
import net.revenj.server.{CommandResult, ReadOnlyServerCommand}

import scala.concurrent.Future
import scala.util.{Failure, Success, Try}

class CountDomainObject(domainModel: DomainModel) extends ReadOnlyServerCommand {

  override def execute[TInput, TOutput](
    locator: ServiceLocator,
    input: Serialization[TInput],
    output: Serialization[TOutput],
    data: TInput): Future[CommandResult[TOutput]] = {

    val arg = input.deserializeRuntime(data, classOf[Argument[TInput]], data.getClass)
    lazy val manifest = domainModel.find(arg.get.Name)
    if (!arg.isSuccess) {
      CommandResult.badRequest(arg.failed)
    } else if (manifest.isEmpty) {
      CommandResult.badRequest(s"Unable to find specified domain object: ${arg.get.Name}")
    } else if (!classOf[DataSource].isAssignableFrom(manifest.get)) {
      CommandResult.badRequest(s"Specified type is not a data source: ${arg.get.Name}")
    } else {
      val tryRepository = locator.resolve(classOf[SearchableRepository[DataSource]], manifest.get)
      if (!tryRepository.isSuccess) {
        CommandResult.badRequest(s"Error resolving repository for: ${arg.get.Name}. Reason: ${tryRepository.failed.get.getMessage}")
      } else {
        val trySpec: Try[Specification[DataSource]] =
          if (arg.get.Specification.isDefined && arg.get.SpecificationName.getOrElse("").nonEmpty) {
            val shortSpec = domainModel.find(arg.get.Name + "+" + arg.get.SpecificationName.get)
            lazy val fullSpec = domainModel.find(arg.get.SpecificationName.get)
            if (shortSpec.isDefined) input.deserializeRuntime[Specification[DataSource]](arg.get.Specification.get, shortSpec.get)
            else if (fullSpec.isDefined) input.deserializeRuntime[Specification[DataSource]](arg.get.Specification.get, fullSpec.get)
            else Failure(new IllegalArgumentException(s"Couldn't find specification: ${arg.get.SpecificationName.get}"))
          } else {
            Success(
              arg.get.Specification match {
                case Some(spec: Specification[DataSource] @unchecked) => spec
                case _ => null
              }
            )
          }
        if (trySpec.isFailure) {
          CommandResult.badRequest(trySpec.failed.get.getMessage)
        } else {
          import scala.concurrent.ExecutionContext.Implicits.global
          tryRepository.get.count(Option(trySpec.get)).map { found =>
            val response = output.serializeRuntime(found)
            if (response.isSuccess) {
              CommandResult[TOutput](Some(response.get), found.toString, 200)
            } else {
              CommandResult[TOutput](None, response.failed.get.getMessage, 500)
            }
          }
        }
      }
    }
  }
}

object CountDomainObject {

  case class Argument[TFormat](
    Name: String,
    SpecificationName: Option[String],
    Specification: Option[TFormat])

}